precompiles/src/secp256k1.rs

```diff
@@ -0,0 +1,406 @@
+            .saturating_add(1);
+        let end = start.saturating_add(SIGNATURE_OFFSETS_SERIALIZED_SIZE);
+
+        let offsets: SecpSignatureOffsets = bincode::deserialize(&data[start..end])
+            .map_err(|_| PrecompileError::InvalidSignature)?;
+
+        // Parse out signature
+        let signature_index = offsets.signature_instruction_index as usize;
+        if signature_index >= instruction_datas.len() {
+            return Err(PrecompileError::InvalidInstructionDataSize);
+        }
+        let signature_instruction = instruction_datas[signature_index];
+        let sig_start = offsets.signature_offset as usize;
+        let sig_end = sig_start.saturating_add(SIGNATURE_SERIALIZED_SIZE);
+        if sig_end >= signature_instruction.len() {
+            return Err(PrecompileError::InvalidSignature);
+        }
+
+        let signature = libsecp256k1::Signature::parse_standard_slice(
+            &signature_instruction[sig_start..sig_end],
+        )
+        .map_err(|_| PrecompileError::InvalidSignature)?;
+
+        let recovery_id = libsecp256k1::RecoveryId::parse(signature_instruction[sig_end])
+            .map_err(|_| PrecompileError::InvalidRecoveryId)?;
+
+        // Parse out pubkey
+        let eth_address_slice = get_data_slice(
+            instruction_datas,
+            offsets.eth_address_instruction_index,
+            offsets.eth_address_offset,
+            HASHED_PUBKEY_SERIALIZED_SIZE,
+        )?;
+
+        // Parse out message
+        let message_slice = get_data_slice(
+            instruction_datas,
+            offsets.message_instruction_index,
+            offsets.message_data_offset,
+            offsets.message_data_size as usize,
+        )?;
+
+        let mut hasher = sha3::Keccak256::new();
+        hasher.update(message_slice);
+        let message_hash = hasher.finalize();
+
+        let pubkey = libsecp256k1::recover(
+            &libsecp256k1::Message::parse_slice(&message_hash).unwrap(),
+            &signature,
+            &recovery_id,
+        )
+        .map_err(|_| PrecompileError::InvalidSignature)?;
+        let eth_address = construct_eth_pubkey(&pubkey);
+
+        if eth_address_slice != eth_address {
+            return Err(PrecompileError::InvalidSignature);
+        }
+    }
+    Ok(())
+}
+
+fn get_data_slice<'a>(
+    instruction_datas: &'a [&[u8]],
+    instruction_index: u8,
+    offset_start: u16,
+    size: usize,
+) -> Result<&'a [u8], PrecompileError> {
+    let signature_index = instruction_index as usize;
+    if signature_index >= instruction_datas.len() {
+        return Err(PrecompileError::InvalidDataOffsets);
+    }
+    let signature_instruction = &instruction_datas[signature_index];
+    let start = offset_start as usize;
+    let end = start.saturating_add(size);
+    if end > signature_instruction.len() {
+        return Err(PrecompileError::InvalidSignature);
+    }
+
+    Ok(&instruction_datas[signature_index][start..end])
+}
```

Comment:
Nothing to do here, but it's a bit baffling that the behavior of this isn't exactly the same between the precompiles

Reply:
> Nothing to do here, but it's a bit baffling that the behavior of this isn't exactly the same between the precompiles

don't look too closely at each of their "offsets" structs